### DISCUSSION TITLE
[feature request] support using tempel-expand via hippie-expand

### DISCUSSION THREAD

**Discussion by pmiam** [minad/tempel#120] [Ideas]
Allow me to make my case or jump to TL;DR

I like to use tempel alongside abbrev as they each do their part in my "second brain"
- tempel saves me from remembering any boilerplate relevant to the language I'm using after setting up the template
- abbrev-mode helps me to quickly absorb and then use acronyms and mathematical expressions in any context
- dabbrev also saves me from typing certain long words repeatedly

my memory sometimes fails me, so having `tempel-insert` to minimize friction when this happens is nice, but usually I prefer to just type the template name and get moving asap. So, having `tempel-complete` as my capf is very convenient.
corfu immediately shows me my options and, with my corfu configuration as below, I can "escape" the pop-up with minimum keystrokes.

(Note, using `tempel-expand` as the capf with this corfu config does not work. Pressing M-TAB uses an eglot completion -- this is probably a mistake in my configuration, but I wanted to mention it)

The first problem with this set up is that tempel-complete is very eager and expands the first matched template immediately. Arguably, this isn't even a problem. Templates are few and usually fairly uniquely named. But I'm sure I could get it to wait for a word-separating character at least.

TL;DR:
Solving the second problem with this set up is currently beyond me. This problem is in the interaction between tempel and dabbrev  (and others) when used through the `hippie-expand` function.

For example, I am working in an elisp-mode buffer with tempel, and tempel-collection

I wish to autotype "logarithm"--  a long word I laboriously entered earlier -- using dabbrev.

So, I start to type "log" (in fact, I type "l" "o" C-g "g" to avoid issue 1) -- the name of the log message template. Then, I go to hippie-expand the rest with `M-/`

If `tempel-expand` supported this paradigm and I prepended it to the `hippie-expand-try-functions-list`, I would be able to expand the elisp log template with the first invocation, then simply press `M-/` again for the dabbrev completion, and as many times as needed again for more distant candidates. Unfortunately, this does not happen.

Currently, you must try hippie-expand twice to get `tempel-expand` to work, and then the function looses track of "log" and denies that further expansions are possible.

This is perhaps also an opportunity to get a `tempel-collapse` function that simply "un-expands" the log message form when `C-u M-/` is typed. This would allow for tempel to place nicely with the hippie-expand undo cannon, but might also be useful to other configurations. In my experience, standard undo backtracks to before the template name was entered.

Thank you, minad and all your contributors very much for these powerful and elegant packages and for your consideration.
Cheers,
Panos

```
(use-package corfu
  :ensure t
  :init
  (global-corfu-mode 1)
  :config
  (setq corfu-auto t
        corfu-auto-prefix 2
        corfu-quit-at-boundary t
        corfu-auto-delay 0
        corfu-cycle t
        corfu-preselect-first nil)
  (unbind-key "RET" corfu-map)
  :hook (shell-mode . (lambda ()
                        (setq-local corfu-quit-at-boundary t
                                    corfu-quit-no-match t
                                    corfu-auto nil))))
					; snippets and templates
(use-package abbrev
  :init
  (setq save-abbrevs 'silently
	abbrev-suggest t)
  :config
  (abbrev-mode 1)
  :bind (("C-x a r" . expand-region-abbrevs)
	 ("C-x a a" . list-abbrevs)
	 ("C-x a w" . edit-abbrevs)
	 ("M-/" . hippie-expand)
	 ("C-x a u" . unexpand-abbrev)))

(use-package tempel
  :ensure t
  :init
  (setq tempel-path (file-name-concat user-emacs-directory
				      "templates/*/*.eld"))
  (defun tempel-setup-capf ()
    "Add Capf to `completion-at-point-functions'. 
    Add `tempel-complete' *before* the mode's main
    Capf, so it will be tried first."
    (setq-local completion-at-point-functions
                (cons #'tempel-complete
                      completion-at-point-functions)))

  (add-hook 'conf-mode-hook 'tempel-setup-capf)
  (add-hook 'prog-mode-hook 'tempel-setup-capf)
  (add-hook 'text-mode-hook 'tempel-setup-capf)
  :config
  (add-to-list 'hippie-expand-try-functions-list
	       'tempel-expand)
  :bind ((:map global-map
               ("C-c y" . tempel-insert))
         (:map tempel-map
               ("M-}" . tempel-next)
               ("M-{" . tempel-previous))))
```

**Comment by minad** [comment]
Hi, thanks for the kind words. I am not sure if I fully understood your question.

1. Regarding eager expansion you may want to try `tempel-trigger-prefix` such that Tempel will only expand when you type the prefix character, e.g., `<`. Alternatively I suggest to trigger Tempel manually, e.g., bind the key `M-+` to `tempel-complete`.
2. I have no experience with the interaction of Hippie/Dabbrev/Tempel. I consider Hippie obsolete and recommend avoiding it, since it predates the Capf completion mechanism. You can instead rely on Dabbrev (e.g. via `cape-dabbrev`), other Capfs (also from Cape or from other packages) and Tempel.

**Reply by pmiam** [reply]
Thank you for your insight. I'll try to re-frame my problem. Essentially, I'm trying to eliminate the need to plan ahead.

With my config `temple-complete` is the first capf to own the pop-up, I'll only get dabbrev/dict/etc if it fails.  I'm trying to manually move to the next capf after I've already started typing. Perhaps in bigger steps than might be achieved by merging capfs together.

**Reply by minad** [reply]
I see. It is a common problem that one may end up with an undesired completion backend. You could try merging the Capfs. Usually I recommend to trigger completion backends manually, see https://github.com/minad/cape#configuration. Only the "main" backend is used for auto completion.

But as an experiment, I just tried the following snippet to cycle through the Capfs. Maybe something like this could be interesting?

~~~ elisp
(setq completion-at-point-functions '(cape-dabbrev cape-dict))

(defun corfu-cycle ()
  (interactive)
  (setq-local completion-at-point-functions
              (nconc (last completion-at-point-functions)
                     (butlast completion-at-point-functions)))
  (completion-at-point))

(keymap-set corfu-map "M-c" #'corfu-cycle)
~~~

**Reply by pmiam** [reply]
This is certainly a superior approach to my initial request. I think I can make this function work for me with some more experimenting.  I will post the configuration I settle on. Thanks for your help!
